core/vm/interpreter.go

```diff
@@ -182,62 +182,56 @@ func (in *EVMInterpreter) Run(contract *Contract, input []byte, readOnly bool) (
 			// Capture pre-execution values for tracing.
 			logged, pcCopy, gasCopy = false, pc, contract.Gas
 		}
-
 		// Get the operation from the jump table and validate the stack to ensure there are
 		// enough stack items available to perform the operation.
 		op = contract.GetOp(pc)
 		operation := in.cfg.JumpTable[op]
+		cost = operation.constantGas // For tracing
```

Comment:
Yesterday I started a branch to experiment with moving all the gas calculation into another function (next to the implementation): https://github.com/ipsilon/go-ethereum/tree/gas-redesign

The constant gas instructions would have the same simple function charging the constant gas, others would have the dynamic calculation + memory expansion.

I am not sure if this will help everything considered, because it will always mean an indirect call, but at the same time it reduces most of the branching in this loop.

Reply:
Looked briefly at the last commit there -- I think it won't work that easily. 

The way the dynamic gas has evolved, we need to first ensure that the caller can cover the static cost, because for some ops, actually _calculating_ the dynamic cost is expensive: checking the existence of things in the trie. 
So we can't just calculate _everything_ and then charge. 

We can get around it by either having the "chargeGas" function not only return the gas, but actually charge it via the `contract`. 
Alternatively, we can pass the availableGas into the "chargeGas" function.